### PR TITLE
feat: resource recorder

### PR DESCRIPTION
**Pull request by chesedo**
## Description of change
Add the new resource recorder as an isolates service

Btw, this used the following command to only extract the resource recorder from the feature branch
```
git cherry-pick $(g rev-list --reverse upstream/feat/shuttle-runtime-scaling -- resource-recorder)
```

## How has this been tested? (if applicable)
With added integration tests

### CODE REVIEW

File: resource-recorder/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+        Ok(resources.into_iter().map(Into::into).collect())
+    }
+
+    /// Delete a resource
+    async fn delete_resource(&self, resource: resource_recorder::Resource) -> Result<(), Error> {
+        self.dal.delete_resource(&resource.try_into()?).await?;
+
+        Ok(())
+    }
```
Comment (iulianbarbu) [inline resource-recorder/src/lib.rs:97]: Yeh, wondering if this is enough. Should we cleanup our infra as well? I would cover this in another PR probably, but the endpoint shouldn't be exposed yet given we remove the resources from the persistence and we might be left with dangling resources in the infra, but maybe I am missing something? Also, we should factor in this endpoint in the custom resources definitions too (and infra that lives outside our management e.g. Turso).

File: resource-recorder/src/type/mod.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+use sqlx::Database;
+
+pub mod database;
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Type {
+    Database(database::Type),
+    Secrets,
+    StaticFolder,
+    Persist,
```
Comment (iulianbarbu) [inline resource-recorder/src/type/mod.rs:12]: I am just wondering if we should have `Turso` here as well. Potentially the custom resource once we merge the #1082 . What do you think?

**Review by iulianbarbu — COMMENTED** [review]
Just a few questions related to Turso and custom resources.

File: resource-recorder/src/lib.rs (hunk 1 above)
Comment (chesedo) [reply resource-recorder/src/lib.rs:97]: Yip, this should not be exposed yet. Whatever will call this (deployer) will have to also call provisioner to cleanup the infra.

File: resource-recorder/src/type/mod.rs (hunk 2 above)
Comment (chesedo) [reply resource-recorder/src/type/mod.rs:12]: Yip, I'll have to add `Turso`

File: resource-recorder/src/lib.rs (hunk 1 above)
Comment (iulianbarbu) [reply resource-recorder/src/lib.rs:97]: Do we still want to expose this RPC (even internally)? Maybe we can add a clarifying comment on the RPC method implementation, that this handler still lacks the infrastructure teardown. What do you think?

**Review by iulianbarbu — APPROVED** [review]

File: resource-recorder/src/lib.rs (hunk 1 above)
Comment (chesedo) [reply resource-recorder/src/lib.rs:97]: Nah, the infrastructure teardown won't go through this RPC anyway (something else will do the teardown and just record it here). And trying to add the comment here feels out of place.

**Pull request merged**